src/system_unix.cpp

```diff
@@ -0,0 +1,214 @@
+struct GetPasswdEntryResult {
+    explicit GetPasswdEntryResult(const std::string& error_) : error(error_) {
+    }
+
+    GetPasswdEntryResult(uid_t uid_, gid_t gid_, const std::vector<gid_t>& groups_) :
+        uid(uid_), gid(gid_), groups(groups_) {
+    }
+
+    std::string error;
+    uid_t uid;
+    gid_t gid;
+    std::vector<gid_t> groups;
+
+    operator bool() const {
+        return this->error.empty();
+    }
+};
+
+static GetPasswdEntryResult get_passwd_entry(const std::string& user_name) {
+    // Assuming that a user does not have more than 50 groups
+    constexpr int max_supplementary_groups = 50;
+
+    const auto entry = getpwnam(user_name.c_str());
+
+    if (not entry) {
+        return GetPasswdEntryResult("Could not get passwd entry for user name: " + user_name);
+    }
+
+    // get supplementary groups for this user
+    int max_ngroups = max_supplementary_groups;
+    gid_t groups[max_supplementary_groups];
+
+    int ngroups = getgrouplist(user_name.c_str(), entry->pw_gid, groups, &max_ngroups);
+    if (ngroups < 0) {
+        return GetPasswdEntryResult("Could not get supplementary groups for user name: " + user_name);
+    }
+
+    return GetPasswdEntryResult(entry->pw_uid, entry->pw_gid, std::vector<gid_t>(groups, groups + ngroups));
+}
+
+bool keep_caps() {
+    return (0 == cap_set_secbits(SECBIT_KEEP_CAPS));
+}
+
+std::string set_caps(const std::vector<std::string>& capabilities) {
+
+    std::vector<cap_value_t> capability_values;
+    capability_values.resize(capabilities.size());
+
+    for (const auto& cap_name : capabilities) {
+        auto& cap_value = capability_values.emplace_back();
+        const auto error = cap_from_name(cap_name.c_str(), &cap_value);
+
+        if (error) {
+            return fmt::format("Failed to get capability value for capability name {}", cap_name);
+        }
+    }
+
+    // FIXME (aw): error handling!
+    auto cap_ctx = cap_get_proc();
+    cap_set_flag(cap_ctx, CAP_INHERITABLE, capability_values.size(), capability_values.data(), CAP_SET);
+    cap_set_proc(cap_ctx);
+    cap_free(cap_ctx);
+
+    for (const auto cap_value : capability_values) {
+        cap_set_ambient(cap_value, CAP_SET);
+    }
+
+    return {};
+}
+
+std::string set_real_user(const std::string& user_name) {
+    // Set special capabilities if required by module
+
+    const auto entry = get_passwd_entry(user_name);
+
+    if (not entry) {
+        return entry.error;
+    }
+
+    const auto set_groups_failed = setgroups(entry.groups.size(), entry.groups.data());
+    if (set_groups_failed) {
+        return "setgroups failed";
+    }
+
+    const auto set_gid_failed = setgid(entry.gid);
+    if (set_gid_failed) {
+        return "setgid failed";
+    }
+
+    const auto set_uid_failed = setuid(entry.uid);
+    if (set_uid_failed) {
+        return "setuid failed";
+    }
+
+    return {};
+}
+
+void SubProcess::send_error_and_exit(const std::string& message) {
+    // FIXME (aw): howto do asserts?
```

Comment:
This will fail when pid == 0 is false, otherwise pass. I think that's correct here

Reply:
removed the fixme